### PR TITLE
Revamp mapping expression context completions

### LINKED ISSUE #31075: Revamp Completions for Mapping Constructor when the resolved type is a Union
**Description:**
$title.
Now, when the context type is resolved as a union, we do not provide any completions for the mapping expression. Although, we need to infer the record types in the union and then infer the fields.

**Steps to reproduce:**
```
type Rec1 record {
    int field1;
    int field2;
};

type Rec2 record {
    int field1;
    int field2;
};

type RecUnion Rec1|Rec2;

public function main() {
    RecUnion x = {<cursor>}
}
```

### PR DESCRIPTION
**Pull request by nadeeshaan**
## Purpose
Revamp mapping expression context completions.

Fixes #31075

https://user-images.githubusercontent.com/1329674/121522377-acaa5c00-ca12-11eb-83d3-1bde4b12eb9e.mp4

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by pubudu91** [comment]
Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?

**Comment by nadeeshaan** [comment]
> Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?

No, with the available fields we can't guarantee that we pick the exact type. There can be situations, where there are same set of fields being a subset of another Record. Therefore, what we do is, when we suggest the fields next time, we omit the already added set of fields.

File: language-server/modules/langserver-core/src/test/resources/completion/annotation_ctx/config/functionAnnotation5.json [hunk 1]
```diff
@@ -1,41 +1 @@
...
-      "label": "Add All Attributes",
-      "kind": "Property",
-      "detail": "none",
-      "sortText": "Q",
-      "insertText": "foo: \"\",\nbar: 0",
-      "insertTextFormat": "Snippet"
-    }
-  ]
-}
+{"position":{"line":3,"character":4},"source":"annotation_ctx/source/functionAnnotation5.bal","items":[{"label":"readonly","kind":"Keyword","detail":"Keyword","sortText":"P","insertText":"readonly ","insertTextFormat":"Snippet"},{"label":"bar","kind":"Field","detail":"(record {| string foo; int bar?; anydata...; |}).bar","sortText":"G","insertText":"bar: ${1:0}","insertTextFormat":"Snippet"},{"label":"foo","kind":"Field","detail":"(record {| string foo; int bar?; anydata...; |}).foo","sortText":"G","insertText":"foo: \"${2}\"","insertTextFormat":"Snippet"},{"label":"Fill record {| string foo; int bar?; anydata...; |} Required Fields","kind":"Property","detail":"record {| string foo; int bar?; anydata...; |}","sortText":"Q","filterText":"fill","insertText":"foo: \"\"","insertTextFormat":"Snippet"}]}
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/resources/completion/annotation_ctx/config/functionAnnotation5.json:1]: would you be able to format this.

File: language-server/modules/langserver-core/src/test/resources/completion/annotation_ctx/config/functionAnnotation6.json [hunk 2]
```diff
@@ -1,41 +1 @@
...
-      "label": "Add All Attributes",
-      "kind": "Property",
-      "detail": "none",
-      "sortText": "Q",
-      "insertText": "foo: \"\",\nbar: 0",
-      "insertTextFormat": "Snippet"
-    }
-  ]
-}
+{"position":{"line":3,"character":5},"source":"annotation_ctx/source/functionAnnotation6.bal","items":[{"label":"readonly","kind":"Keyword","detail":"Keyword","sortText":"P","insertText":"readonly ","insertTextFormat":"Snippet"},{"label":"bar","kind":"Field","detail":"(record {| string foo; int bar?; anydata...; |}).bar","sortText":"G","insertText":"bar: ${1:0}","insertTextFormat":"Snippet"},{"label":"foo","kind":"Field","detail":"(record {| string foo; int bar?; anydata...; |}).foo","sortText":"G","insertText":"foo: \"${2}\"","insertTextFormat":"Snippet"},{"label":"Fill record {| string foo; int bar?; anydata...; |} Required Fields","kind":"Property","detail":"record {| string foo; int bar?; anydata...; |}","sortText":"Q","filterText":"fill","insertText":"foo: \"\"","insertTextFormat":"Snippet"}]}
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/resources/completion/annotation_ctx/config/functionAnnotation6.json:1]: Shall we format this

**Review by malinthar — APPROVED** [review]

**Comment by pubudu91** [comment]
> > Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?
> 
> No, with the available fields we can't guarantee that we pick the exact type. There can be situations, where there are same set of fields being a subset of another Record. Therefore, what we do is, when we suggest the fields next time, we omit the already added set of fields.

Agree about multiple record types having the same fields (names and types). But after we finish filling those common ones, and then pick a field unique to a particular type, wouldn't it be sorta misleading to keep suggesting the fields in others types?

**Comment by nadeeshaan** [comment]
> > > Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?
> > 
> > 
> > No, with the available fields we can't guarantee that we pick the exact type. There can be situations, where there are same set of fields being a subset of another Record. Therefore, what we do is, when we suggest the fields next time, we omit the already added set of fields.
> 
> Agree about multiple record types having the same fields (names and types). But after we finish filling those common ones, and then pick a field unique to a particular type, wouldn't it be sorta misleading to keep suggesting the fields in others types?

Got your point. We should handle that. Let me send a revamped version.

**Comment by nadeeshaan** [comment]
> > > Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?
> > 
> > 
> > No, with the available fields we can't guarantee that we pick the exact type. There can be situations, where there are same set of fields being a subset of another Record. Therefore, what we do is, when we suggest the fields next time, we omit the already added set of fields.
> 
> Agree about multiple record types having the same fields (names and types). But after we finish filling those common ones, and then pick a field unique to a particular type, wouldn't it be sorta misleading to keep suggesting the fields in others types?

Consider the following use-case

type Record1 record {

}

> > > Suppose I pick one of the fields from the completion list and give it a value. Then, I go ahead and check the completions again. What would the completions list look like in that case? Does it determine the type of the mapping constructor based on that one field I filled?
> > 
> > 
> > No, with the available fields we can't guarantee that we pick the exact type. There can be situations, where there are same set of fields being a subset of another Record. Therefore, what we do is, when we suggest the fields next time, we omit the already added set of fields.
> 
> Agree about multiple record types having the same fields (names and types). But after we finish filling those common ones, and then pick a field unique to a particular type, wouldn't it be sorta misleading to keep suggesting the fields in others types?

Got it. With the current implementation, we cannot consistently check whether the mapping expression's type is a given TypeSymbol. Hence we had to go ahead with this implementation.

**Review by mohanvive — APPROVED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingConstructorExpressionNodeContext.java [hunk 3]
```diff
@@ -212,5 +214,20 @@ private boolean hasReadonlyKW(NonTerminalNode evalNodeAtCursor) {
         return this.getCompletionItemList(moduleContent, context);
     }
 
+    private Map<String, RecordFieldSymbol> getValidFields(MappingConstructorExpressionNode node,
+                                                          RecordTypeSymbol recordTypeSymbol) {
+        List<String> missingFields = node.fields().stream()
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingConstructorExpressionNodeContext.java:219]: Shouldn't this variable name be `alreadyAddedFields`?

**Review by malinthar — APPROVED** [review]

**Review by IMS94 — APPROVED** [review]
LGTM! There's a minor comment about a variable name.

**Pull request merged**